### PR TITLE
Sync updated protocol test cases and apply necessary fixes

### PR DESCRIPTION
**Pull request by jeskew**
Fix parser tests

Fix serializer tests

Use locationName of shapes even when they are flattened lists whose members have no locationName

### CODE REVIEW

File: src/Api/Parser/AbstractRestParser.php [hunk 1]
```diff
@@ -89,18 +89,30 @@ private function extractHeader(
...
-                $value = new DateTimeResult($value);
-            } catch (\Exception $e) {
-                // If the value cannot be parsed, then do not add it to the
-                // output structure.
-                return;
-            }
+        switch ($shape->getType()) {
+            case 'float':
+            case 'double':
+                $value = doubleval($value);
```
Comment (GrahamCampbell) [inline src/Api/Parser/AbstractRestParser.php]: Using `(double) $value` is faster and is identical in terms of the casting behaviour.

File: src/Api/Parser/AbstractRestParser.php [hunk 2]
```diff
@@ -89,18 +89,30 @@ private function extractHeader(
...
+                break;
+            case 'long':
+                $value = intval($value);
```
Comment (GrahamCampbell) [inline src/Api/Parser/AbstractRestParser.php]: `(int) $value`, etc

File: src/Api/Parser/AbstractRestParser.php (hunk 1 above)
Comment (GrahamCampbell) [inline src/Api/Parser/AbstractRestParser.php]: Or rather, `(float) $value`.

Comment (mtdowling) [inline src/Api/Parser/AbstractRestParser.php]: :+1:  Agreed, let's use casts instead

**Comment by jeskew** [comment]
Addresses (among other things) #687

**Comment by mtdowling** [comment]
:shipit:

**Pull request merged**